// feel-commander/dist/commands/core/install.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
const fsExtra = tslib_1.__importStar(require("fs-extra"));
const listr_1 = tslib_1.__importDefault(require("listr"));
const os = tslib_1.__importStar(require("os"));
const base_1 = tslib_1.__importDefault(require("../../base"));
const constants_1 = require("../../utils/constants");
const commons_1 = require("../../utils/core/commons");
const database_1 = require("../../utils/core/database");
const pm2_1 = require("../../utils/core/pm2");
const release_1 = require("../../utils/core/release");
const download_1 = require("../../utils/download");
const flags_1 = require("../../utils/flags");
const start_1 = tslib_1.__importDefault(require("./start"));
const validatePrerequisite = (installPath) => {
    if (!commons_1.isSupportedOS()) {
        throw new Error(`Feel Core installation is not supported on ${os.type()}.`);
    }
    if (fsExtra.pathExistsSync(installPath)) {
        throw new Error(`Feel Core installation already exists in path ${installPath}.`);
    }
};
const validateFlags = ({ network, 'release-url': releaseUrl, 'snapshot-url': snapshotUrl, }) => {
    commons_1.validateNetwork(network);
    if (releaseUrl) {
        commons_1.validURL(releaseUrl);
    }
    if (snapshotUrl) {
        commons_1.validURL(snapshotUrl);
    }
};
const installOptions = async ({ 'installation-path': installationPath, network, 'release-url': releaseUrl, 'feel-version': feelVersion, }, name) => {
    const installPath = commons_1.feelInstall(installationPath);
    const installDir = `${installPath}/${name}/`;
    const latestUrl = releaseUrl || commons_1.feelLatestUrl(constants_1.RELEASE_URL, network);
    const installVersion = await commons_1.getVersionToInstall(network, feelVersion, releaseUrl);
    const { version, feelTarUrl, feelTarSHA256Url } = await release_1.getReleaseInfo(latestUrl, network, installVersion);
    return {
        installDir,
        version,
        feelTarUrl,
        feelTarSHA256Url,
        latestUrl,
    };
};
class InstallCommand extends base_1.default {
    async run() {
        const { args, flags } = this.parse(InstallCommand);
        const { 'feel-version': feelVersion, 'no-snapshot': noSnapshot, 'no-start': noStart, network, 'snapshot-url': snapshotUrl, } = flags;
        const { name } = args;
        const { cacheDir } = this.config;
        fsExtra.ensureDirSync(cacheDir);
        const snapshotURL = commons_1.feelSnapshotUrl(snapshotUrl, network);
        const tasks = new listr_1.default([
            {
                title: `Install Feel Core ${network} instance as ${name}`,
                task: () => new listr_1.default([
                    {
                        title: 'Prepare Install Options',
                        task: async (ctx) => {
                            const options = await installOptions(flags, name);
                            ctx.options = options;
                        },
                    },
                    {
                        title: 'Validate root user, flags, prerequisites',
                        task: async (ctx) => {
                            const { installDir, latestUrl } = ctx.options;
                            commons_1.validateNotARootUser();
                            validateFlags(flags);
                            validatePrerequisite(installDir);
                            if (feelVersion) {
                                await commons_1.validateVersion(latestUrl, feelVersion);
                                ctx.options.version = feelVersion;
                            }
                        },
                    },
                    {
                        title: 'Download Feel Core Release and Blockchain Snapshot',
                        task: async (ctx) => {
                            const { feelTarUrl } = ctx.options;
                            if (!noSnapshot && snapshotURL.trim() !== '') {
                                await download_1.download(snapshotURL, cacheDir);
                            }
                            await download_1.downloadAndValidate(feelTarUrl, cacheDir);
                        },
                    },
                    {
                        title: 'Extract Feel Core',
                        task: async (ctx) => {
                            const { installDir, feelTarUrl } = ctx.options;
                            const { fileName, fileDir } = commons_1.getDownloadedFileInfo(feelTarUrl, cacheDir);
                            commons_1.createDirectory(installDir);
                            await download_1.extract(fileDir, fileName, installDir);
                        },
                    },
                    {
                        title: 'Register Feel Core',
                        task: async (ctx) => {
                            const { installDir } = ctx.options;
                            const envConfig = await commons_1.generateEnvConfig(network);
                            await pm2_1.registerApplication(installDir, network, name, envConfig);
                        },
                    },
                    {
                        title: 'Create Database and restore Feel Blockchain Snapshot',
                        task: async (ctx) => {
                            const { installDir } = ctx.options;
                            await database_1.initDB(installDir);
                            await database_1.startDatabase(installDir, name);
                            await database_1.createUser(installDir, network, name);
                            await database_1.createDatabase(installDir, network, name);
                            if (!noSnapshot) {
                                const { filePath } = commons_1.getDownloadedFileInfo(snapshotURL, cacheDir);
                                await database_1.restoreSnapshot(installDir, network, filePath, name);
                            }
                            await database_1.stopDatabase(installDir, name);
                        },
                    },
                ]),
            },
        ]);
        try {
            const instance = await pm2_1.describeApplication(name);
            if (instance) {
                this.log(`\n Feel Core instance ${name} already installed: `);
                this.print(instance);
                return;
            }
            await tasks.run();
            if (!noStart) {
                await start_1.default.run([name]);
                const newInstance = await pm2_1.describeApplication(name);
                this.print(newInstance);
                return;
            }
        }
        catch (error) {
            this.error(JSON.stringify(error));
            const { installDir } = error.context.options;
            const dirPath = installDir.substr(0, installDir.length - 1);
            fsExtra.emptyDirSync(installDir);
            fsExtra.rmdirSync(dirPath);
        }
    }
}
InstallCommand.args = [
    {
        name: 'name',
        description: 'Feel Core installation directory name.',
        required: true,
    },
];
InstallCommand.description = 'Install an instance of Feel Core.';
InstallCommand.examples = [
    'core:install feel-mainnet',
    'core:install --no-start feel-mainnet',
    'core:install --no-snapshot feel-mainnet',
    'core:install --feel-version=2.0.0 feel-mainnet',
    'core:install --network=testnet --release-url=https://downloads.feel.surf/feel/mainnet/1.6.0/feel-1.6.0-Linux-x86_64.tar.gz feel-mainnet',
    'core:install --network=mainnet --snapshot-url=https://downloads.feel.surf/feel/mainnet/blockchain.db.gz custom-mainnet',
];
InstallCommand.flags = {
    json: command_1.flags.boolean(Object.assign({}, base_1.default.flags.json, { hidden: true })),
    pretty: command_1.flags.boolean(Object.assign({}, base_1.default.flags.pretty, { hidden: true })),
    'installation-path': command_1.flags.string(Object.assign({}, flags_1.flags.installationPath, { default: '~/.feel/instances', hidden: true })),
    'feel-version': command_1.flags.string(Object.assign({}, flags_1.flags.feelVersion)),
    'no-snapshot': command_1.flags.boolean(Object.assign({}, flags_1.flags.noSnapshot, { default: false, allowNo: false })),
    'no-start': command_1.flags.boolean(Object.assign({}, flags_1.flags.noStart, { default: false, allowNo: false })),
    network: command_1.flags.string(Object.assign({}, flags_1.flags.network, { default: constants_1.NETWORK.MAINNET, options: [
            constants_1.NETWORK.MAINNET,
            constants_1.NETWORK.TESTNET,
            constants_1.NETWORK.BETANET,
            constants_1.NETWORK.ALPHANET,
            constants_1.NETWORK.DEVNET,
        ] })),
    'release-url': command_1.flags.string(Object.assign({}, flags_1.flags.releaseUrl)),
    'snapshot-url': command_1.flags.string(Object.assign({}, flags_1.flags.snapshotUrl, { default: constants_1.SNAPSHOT_URL })),
};
exports.default = InstallCommand;
