// feel-commander/dist/commands/core/upgrade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
const fsExtra = tslib_1.__importStar(require("fs-extra"));
const listr_1 = tslib_1.__importDefault(require("listr"));
const semver_1 = tslib_1.__importDefault(require("semver"));
const base_1 = tslib_1.__importDefault(require("../../base"));
const constants_1 = require("../../utils/constants");
const cache_1 = require("../../utils/core/cache");
const commons_1 = require("../../utils/core/commons");
const database_1 = require("../../utils/core/database");
const pm2_1 = require("../../utils/core/pm2");
const release_1 = require("../../utils/core/release");
const download_1 = require("../../utils/download");
const flags_1 = require("../../utils/flags");
class UpgradeCommand extends base_1.default {
    async run() {
        const { args, flags } = this.parse(UpgradeCommand);
        const { name } = args;
        const { 'feel-version': feelVersion, 'release-url': releaseUrl, } = flags;
        const instance = await pm2_1.describeApplication(name);
        if (!instance) {
            throw new Error(`Feel Core instance: ${name} doesn't exists.\nTo upgrade first install using feel core:install and then run feel core:upgrade`);
        }
        const { installationPath, network, version: currentVersion, } = instance;
        const upgradeVersion = await commons_1.getVersionToInstall(network, feelVersion, releaseUrl);
        const { cacheDir } = this.config;
        fsExtra.ensureDirSync(cacheDir);
        const tasks = new listr_1.default([
            {
                title: 'Validate Version Input',
                task: async () => {
                    if (semver_1.default.lte(upgradeVersion, currentVersion)) {
                        throw new Error(`Upgrade version:${upgradeVersion} should be greater than current version: ${currentVersion}`);
                    }
                    const latestUrl = releaseUrl || commons_1.feelLatestUrl(constants_1.RELEASE_URL, network);
                    await commons_1.validateVersion(latestUrl, upgradeVersion);
                },
            },
            {
                title: `Download Feel Core: ${upgradeVersion} for upgrade`,
                task: async () => {
                    const { feelTarUrl } = await release_1.getReleaseInfo(releaseUrl, network, upgradeVersion);
                    await download_1.downloadAndValidate(feelTarUrl, cacheDir);
                },
            },
            {
                title: 'Stop, Backup and Install Feel Core',
                task: () => new listr_1.default([
                    {
                        title: `Stop Feel Core`,
                        task: async () => {
                            const isRunning = await cache_1.isCacheRunning(installationPath, name);
                            if (isRunning) {
                                await cache_1.stopCache(installationPath, network, name);
                            }
                            await database_1.stopDatabase(installationPath, name);
                            await pm2_1.stopApplication(name);
                        },
                    },
                    {
                        title: `Backup Feel Core: ${currentVersion} installed as ${name}`,
                        task: () => {
                            commons_1.backupFeel(installationPath, name);
                        },
                    },
                    {
                        title: `Install Feel Core: ${upgradeVersion}`,
                        task: async () => {
                            fsExtra.ensureDirSync(installationPath);
                            const { feelTarUrl } = await release_1.getReleaseInfo(releaseUrl, network, upgradeVersion);
                            const { fileDir, fileName } = commons_1.getDownloadedFileInfo(feelTarUrl, cacheDir);
                            await download_1.extract(fileDir, fileName, installationPath);
                        },
                    },
                ]),
            },
            {
                title: `Upgrade Feel Core ${name} instance from: ${currentVersion} to: ${upgradeVersion}`,
                task: async () => {
                    await commons_1.upgradeFeel(installationPath, name, network, currentVersion);
                },
            },
            {
                title: 'Unregister and register Feel Core',
                task: async () => {
                    const envConfig = await commons_1.generateEnvConfig(network);
                    await pm2_1.unRegisterApplication(name);
                    await pm2_1.registerApplication(installationPath, network, name, envConfig);
                },
            },
            {
                title: `Start Feel Core: ${upgradeVersion}`,
                task: async () => {
                    const isRunning = await cache_1.isCacheRunning(installationPath, name);
                    if (!isRunning) {
                        await cache_1.startCache(installationPath, name);
                    }
                    await database_1.startDatabase(installationPath, name);
                    await pm2_1.restartApplication(name);
                },
            },
        ]);
        await tasks.run();
    }
}
UpgradeCommand.args = [
    {
        name: 'name',
        description: 'Feel Core installation directory name.',
        required: true,
    },
];
UpgradeCommand.description = 'Upgrade an instance of Feel Core to a specified or latest version.';
UpgradeCommand.examples = [
    'core:upgrade feel-mainnet',
    'core:upgrade --feel-version=2.0.0 feel-mainnet',
    'core:upgrade --release-url=https://downloads.feel.surf/feel/mainnet/2.1.0/feel-2.1.0-Linux-x86_64.tar.gz feel-mainnet',
];
UpgradeCommand.flags = {
    json: command_1.flags.boolean(Object.assign({}, base_1.default.flags.json, { hidden: true })),
    pretty: command_1.flags.boolean(Object.assign({}, base_1.default.flags.pretty, { hidden: true })),
    'feel-version': command_1.flags.string(Object.assign({}, flags_1.flags.feelVersion)),
    'release-url': command_1.flags.string(Object.assign({}, flags_1.flags.releaseUrl)),
};
exports.default = UpgradeCommand;
