// feel-commander/dist/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const feel_constants_1 = require("@feelhq/feel-constants");
exports.COMMAND_TYPES = [
    'accounts',
    'addresses',
    'blocks',
    'delegates',
    'transactions',
];
exports.PLURALS = {
    account: 'accounts',
    address: 'addresses',
    block: 'blocks',
    delegate: 'delegates',
    transaction: 'transactions',
};
exports.QUERY_INPUT_MAP = {
    accounts: 'address',
    blocks: 'blockId',
    delegates: 'username',
    transactions: 'id',
};
exports.CONFIG_VARIABLES = [
    'api.nodes',
    'api.network',
    'json',
    'pretty',
];
exports.API_PROTOCOLS = ['http:', 'https:'];
exports.NETHASHES = {
    main: feel_constants_1.MAINNET_NETHASH,
    test: feel_constants_1.TESTNET_NETHASH,
};
exports.SORT_FIELDS = [
    'publicKey:asc',
    'publicKey:desc',
    'balance:asc',
    'balance:desc',
    'username:asc',
    'username:desc',
];
var NETWORK;
(function (NETWORK) {
    NETWORK["MAINNET"] = "mainnet";
    NETWORK["TESTNET"] = "testnet";
    NETWORK["BETANET"] = "betanet";
    NETWORK["ALPHANET"] = "alphanet";
    NETWORK["DEVNET"] = "devnet";
})(NETWORK = exports.NETWORK || (exports.NETWORK = {}));
exports.POSTGRES_PORT = 5432;
exports.REDIS_PORT = 6379;
exports.HTTP_PORTS = {
    mainnet: 8500,
    testnet: 7500,
    betanet: 6800,
    alphanet: 6500,
    devnet: 4500,
};
exports.WS_PORTS = {
    mainnet: 8501,
    testnet: 7501,
    betanet: 6801,
    alphanet: 6501,
    devnet: 4001,
};
var OS;
(function (OS) {
    OS["Darwin"] = "MACOS";
    OS["Linux"] = "LINUX";
})(OS = exports.OS || (exports.OS = {}));
exports.RELEASE_URL = 'https://downloads.feel.surf/feel';
exports.SNAPSHOT_URL = 'https://downloads.feel.surf/feel/mainnet/blockchain.db.gz';
